### PR TITLE
Allow load/save a model from/to an existing HDF5 file.

### PR DESCRIPTION
**Pull request by jorgecarleitao**
The changes are minimal but the diff is large because there is a piece of code that was indented one level down.

Apart from this, I just re-wrote a function that 
1. opens a file
2. writes\reads on it

into a function that only does 1 if the argument is a string.

The public interface is **backward compatible**: the user can pass either an HDF file descriptor or a string (to create the file). This is in line e.g. with the interface used by pandas.

In essence, the (tested) feature is that we can now use

```python
    with h5py.File(fname, 'w') as h5file:
        group = h5file.create_group('model')
        save_model(model, group)

    with h5py.File(fname, 'r') as h5file:
        model = load_model(h5file['model'])
```

to save a model.

### CODE REVIEW

File: tests/test_model_saving.py [hunk 1]
```diff
@@ -102,6 +103,34 @@ def test_functional_model_saving():
...
+                  optimizer=optimizers.RMSprop(lr=0.0001),
+                  metrics=[metrics.categorical_accuracy])
+    x = np.random.random((1, 3))
+    y = np.random.random((1, 3))
+    model.train_on_batch(x, y)
+
+    out = model.predict(x)
+    _, fname = tempfile.mkstemp('.h5')
+    with h5py.File(fname, 'w') as h5file:
+        group = h5file.create_group('model')
```
Comment (Dapid) [inline tests/test_model_saving.py]: Can you save something else in the file as well? Random garbage, and maybe another model, to make sure we are loading the correct one.

**Comment by fchollet** [comment]
Why do you need both a `_load_model` (used once) and `load_model` method? Why do you need both `_save_model` (used once) and `save_model` method? Wouldn't it make more sense to have a single function for each?

**Comment by jorgecarleitao** [comment]
@fchollet, two reasons:

1. AFAIK it is not possible to write a conditional of a `with` block in Python. Specifically, how would you re-write this piece of code without `_load_model`?

```
if isinstance(filepath_or_f, str):
     with h5py.File(filepath_or_f, mode='r') as f:
         model = _load_model(f, custom_objects, compile)
else:
     model = _load_model(filepath_or_f, custom_objects, compile)
```
(each `_*_model` is used twice in the respective `_*_model`)

2. So that different functionality is separated in the code. The `_save_model` implements how we save a model into HDF, the `save_model` implements how we expose this functionality. IMO it makes both `*_model` and `_*_model` more readable. If we modify any of them, we easily know what we will need to inform the user on: the user-saved models (serialization, `_save_model`) or the user-written code (API, `save_model`).

**Comment by jorgecarleitao** [comment]
@Dapid, Thanks for the review and comment. I modified the test to include loading/unloading of an array into the h5 to test the expected behavior.

**Comment by jorgecarleitao** [comment]
This meeds merging work and therefore will be closed until it is ready.

**Pull request closed without merging**